File: tests/benchmarks/fixtures/Sample.kt
// Synthetic A/B fixture — Kotlin key-value store with TTL and LRU eviction.
// NOT for production use. Generated to exercise tree-sitter outline.

package store

import java.time.Instant
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write
import kotlin.math.min

data class StoreEntry<V>(
    val value: V,
    val insertedAt: Instant,
    val ttlSeconds: Long?,
    var accessCount: Long = 0,
    var lastAccessed: Instant = Instant.now(),
)

data class StoreStats(
    val size: Int,
    val hits: Long,
    val misses: Long,
    val evictions: Long,
    val expirations: Long,
)

class StoreException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)

interface KVStore<K, V> {
    fun get(key: K): V?
    fun put(key: K, value: V, ttlSeconds: Long? = null)
    fun remove(key: K): Boolean
    fun contains(key: K): Boolean
    fun size(): Int
    fun clear()
    fun stats(): StoreStats
}

class LruTtlStore<K, V>(
    private val capacity: Int = 1024,
    private val defaultTtlSeconds: Long? = null,
) : KVStore<K, V> {

    private val map = LinkedHashMap<K, StoreEntry<V>>(capacity, 0.75f, true)
    private val lock = ReentrantReadWriteLock()
    private var hits = 0L
    private var misses = 0L
    private var evictions = 0L
    private var expirations = 0L

    override fun get(key: K): V? = lock.read {
        val entry = map[key] ?: run { misses++; return@read null }
        val ttl = entry.ttlSeconds ?: defaultTtlSeconds
        if (ttl != null && entry.insertedAt.plusSeconds(ttl).isBefore(Instant.now())) {
            lock.write { map.remove(key); expirations++ }
            misses++
            return@read null
        }
        entry.accessCount++
        entry.lastAccessed = Instant.now()
        hits++
        entry.value
    }

    override fun put(key: K, value: V, ttlSeconds: Long?) = lock.write {
        if (map.size >= capacity && !map.containsKey(key)) {
            val oldest = map.keys.first()
            map.remove(oldest)
            evictions++
        }
        map[key] = StoreEntry(
            value = value,
            insertedAt = Instant.now(),
            ttlSeconds = ttlSeconds ?: defaultTtlSeconds,
        )
    }

    override fun remove(key: K): Boolean = lock.write { map.remove(key) != null }

    override fun contains(key: K): Boolean = lock.read { map.containsKey(key) }

    override fun size(): Int = lock.read { map.size }

    override fun clear() = lock.write { map.clear() }

    override fun stats(): StoreStats = lock.read {
        StoreStats(
            size = map.size,
            hits = hits,
            misses = misses,
            evictions = evictions,
            expirations = expirations,
        )
    }

    fun purgeExpired(): Int = lock.write {
        val now = Instant.now()
        val before = map.size
        val toRemove = map.entries.filter { (_, e) ->
            val ttl = e.ttlSeconds ?: defaultTtlSeconds
            ttl != null && e.insertedAt.plusSeconds(ttl).isBefore(now)
        }.map { it.key }
        toRemove.forEach { map.remove(it); expirations++ }
        before - map.size
    }

    fun hotKeys(limit: Int = 10): List<K> = lock.read {
        map.entries
            .sortedByDescending { it.value.accessCount }
            .take(min(limit, map.size))
            .map { it.key }
    }

    fun snapshot(): Map<K, V> = lock.read {
        map.entries.associate { (k, e) -> k to e.value }
    }
}

class NamespacedStore<K, V>(
    private val delegate: KVStore<K, V>,
    private val namespace: String,
    private val separator: String = ":",
) {
    @Suppress("UNCHECKED_CAST")
    private fun ns(key: K): K = "$namespace$separator$key" as K

    fun get(key: K): V? = delegate.get(ns(key))
    fun put(key: K, value: V, ttlSeconds: Long? = null) = delegate.put(ns(key), value, ttlSeconds)
    fun remove(key: K): Boolean = delegate.remove(ns(key))
    fun contains(key: K): Boolean = delegate.contains(ns(key))
}

object StoreFactory {
    fun <K, V> lruTtl(
        capacity: Int = 1024,
        defaultTtlSeconds: Long? = null,
    ): LruTtlStore<K, V> = LruTtlStore(capacity, defaultTtlSeconds)

    fun <K, V> namespaced(
        namespace: String,
        capacity: Int = 1024,
    ): NamespacedStore<K, V> {
        val inner = LruTtlStore<K, V>(capacity)
        return NamespacedStore(inner, namespace)
    }
}

fun <K, V> KVStore<K, V>.getOrPut(key: K, ttlSeconds: Long? = null, block: () -> V): V {
    return get(key) ?: block().also { put(key, it, ttlSeconds) }
}

fun <K, V> KVStore<K, V>.getOrThrow(key: K): V {
    return get(key) ?: throw StoreException("Key not found: $key")
}

fun <K, V> KVStore<K, V>.putAll(entries: Map<K, V>, ttlSeconds: Long? = null) {
    entries.forEach { (k, v) -> put(k, v, ttlSeconds) }
    }

    fun <K, V> KVStore<K, V>.computeIfAbsent(key: K, ttlSeconds: Long? = null, compute: (K) -> V): V {
    return get(key) ?: compute(key).also { put(key, it, ttlSeconds) }
    }

    // ---------- Metrics store ----------

    data class MetricPoint(
    val name: String,
    val value: Double,
    val tags: Map<String, String> = emptyMap(),
    val timestamp: Instant = Instant.now(),
    )

    interface MetricSink {
    fun record(point: MetricPoint)
    fun flush(): List<MetricPoint>
    }

    class InMemoryMetricSink : MetricSink {
    private val buffer = mutableListOf<MetricPoint>()
    private val lock = ReentrantReadWriteLock()

    override fun record(point: MetricPoint) = lock.write { buffer.add(point) }

    override fun flush(): List<MetricPoint> = lock.write {
        val copy = buffer.toList()
        buffer.clear()
        copy
    }

    fun size(): Int = lock.read { buffer.size }
    }

    class TaggedMetricSink(
    private val delegate: MetricSink,
    private val baseTags: Map<String, String>,
    ) : MetricSink {
    override fun record(point: MetricPoint) {
        delegate.record(point.copy(tags = baseTags + point.tags))
    }

    override fun flush(): List<MetricPoint> = delegate.flush()
    }

    class SamplingMetricSink(
    private val delegate: MetricSink,
    private val sampleRate: Double = 0.1,
    ) : MetricSink {
    override fun record(point: MetricPoint) {
        if (Math.random() < sampleRate) delegate.record(point)
    }

    override fun flush(): List<MetricPoint> = delegate.flush()
    }

    // ---------- Store metrics integration ----------

    fun <K, V> LruTtlStore<K, V>.recordMetrics(sink: MetricSink, name: String) {
    val s = stats()
    sink.record(MetricPoint("$name.hits", s.hits.toDouble()))
    sink.record(MetricPoint("$name.misses", s.misses.toDouble()))
    sink.record(MetricPoint("$name.evictions", s.evictions.toDouble()))
    sink.record(MetricPoint("$name.size", s.size.toDouble()))
    }

    object MetricSinkFactory {
    fun inMemory(): InMemoryMetricSink = InMemoryMetricSink()
    fun tagged(delegate: MetricSink, vararg tags: Pair<String, String>): TaggedMetricSink =
        TaggedMetricSink(delegate, mapOf(*tags))
    fun sampling(delegate: MetricSink, rate: Double = 0.1): SamplingMetricSink =
        SamplingMetricSink(delegate, rate)
    }

File: tests/benchmarks/fixtures/sample.swift
// Synthetic A/B fixture — Swift async HTTP cache layer.
// NOT for production use. Generated to exercise tree-sitter outline.

import Foundation

// MARK: - Errors

public enum CacheError: Error, LocalizedError {
    case keyNotFound(String)
    case encodingFailed(Error)
    case decodingFailed(Error)
    case storageFull(capacity: Int)

    public var errorDescription: String? {
        switch self {
        case .keyNotFound(let k): return "Key not found: \(k)"
        case .encodingFailed(let e): return "Encoding failed: \(e.localizedDescription)"
        case .decodingFailed(let e): return "Decoding failed: \(e.localizedDescription)"
        case .storageFull(let cap): return "Cache is full (capacity: \(cap))"
        }
    }
}

// MARK: - Protocol

public protocol AsyncCache<Value>: Sendable {
    associatedtype Value: Codable & Sendable
    func get(forKey key: String) async throws -> Value
    func set(_ value: Value, forKey key: String, ttl: TimeInterval?) async throws
    func delete(forKey key: String) async throws -> Bool
    func flush() async
    func stats() async -> CacheStats
}

public struct CacheStats: Sendable {
    public var hits: Int
    public var misses: Int
    public var size: Int
    public var evictions: Int

    public var hitRate: Double {
        let total = hits + misses
        guard total > 0 else { return 0 }
        return Double(hits) / Double(total)
    }
}

// MARK: - Entry

private struct CacheEntry<V: Codable>: Codable {
    let value: V
    let insertedAt: Date
    let ttl: TimeInterval?
    var accessCount: Int

    var isExpired: Bool {
        guard let ttl else { return false }
        return Date().timeIntervalSince(insertedAt) > ttl
    }
}

// MARK: - In-Memory Implementation

public actor MemoryCache<V: Codable & Sendable>: AsyncCache {
    public typealias Value = V

    private var store: [String: CacheEntry<V>] = [:]
    private var insertionOrder: [String] = []
    private let capacity: Int
    private var stats_: CacheStats = CacheStats(hits: 0, misses: 0, size: 0, evictions: 0)

    public init(capacity: Int = 256) {
        self.capacity = capacity
    }

    public func get(forKey key: String) async throws -> V {
        guard var entry = store[key] else {
            stats_.misses += 1
            throw CacheError.keyNotFound(key)
        }
        if entry.isExpired {
            store.removeValue(forKey: key)
            insertionOrder.removeAll { $0 == key }
            stats_.misses += 1
            throw CacheError.keyNotFound(key)
        }
        entry.accessCount += 1
        store[key] = entry
        stats_.hits += 1
        return entry.value
    }

    public func set(_ value: V, forKey key: String, ttl: TimeInterval? = nil) async throws {
        if store[key] == nil && store.count >= capacity {
            evictLRU()
        }
        if store[key] == nil {
            insertionOrder.append(key)
        }
        store[key] = CacheEntry(value: value, insertedAt: Date(), ttl: ttl, accessCount: 0)
        stats_.size = store.count
    }

    public func delete(forKey key: String) async throws -> Bool {
        guard store.removeValue(forKey: key) != nil else { return false }
        insertionOrder.removeAll { $0 == key }
        stats_.size = store.count
        return true
    }

    public func flush() async {
        store.removeAll()
        insertionOrder.removeAll()
        stats_.size = 0
    }

    public func stats() async -> CacheStats {
        return stats_
    }

    public func purgeExpired() async -> Int {
        let before = store.count
        let expired = store.filter { $0.value.isExpired }.map { $0.key }
        expired.forEach { key in
            store.removeValue(forKey: key)
            insertionOrder.removeAll { $0 == key }
        }
        stats_.size = store.count
        return before - store.count
    }

    private func evictLRU() {
        guard let oldest = insertionOrder.first else { return }
        store.removeValue(forKey: oldest)
        insertionOrder.removeFirst()
        stats_.evictions += 1
    }
}

// MARK: - Layered Cache

public actor LayeredCache<V: Codable & Sendable>: AsyncCache {
    public typealias Value = V

    private let l1: MemoryCache<V>
    private let l2: MemoryCache<V>

    public init(l1Capacity: Int = 64, l2Capacity: Int = 512) {
        l1 = MemoryCache(capacity: l1Capacity)
        l2 = MemoryCache(capacity: l2Capacity)
    }

    public func get(forKey key: String) async throws -> V {
        if let v = try? await l1.get(forKey: key) { return v }
        let v = try await l2.get(forKey: key)
        try await l1.set(v, forKey: key)
        return v
    }

    public func set(_ value: V, forKey key: String, ttl: TimeInterval? = nil) async throws {
        try await l1.set(value, forKey: key, ttl: ttl)
        try await l2.set(value, forKey: key, ttl: ttl)
    }

    public func delete(forKey key: String) async throws -> Bool {
        let r1 = try await l1.delete(forKey: key)
        let r2 = try await l2.delete(forKey: key)
        return r1 || r2
    }

    public func flush() async {
        await l1.flush()
        await l2.flush()
    }

    public func stats() async -> CacheStats {
        let s1 = await l1.stats()
        let s2 = await l2.stats()
        return CacheStats(
            hits: s1.hits + s2.hits,
            misses: s1.misses + s2.misses,
            size: s1.size + s2.size,
            evictions: s1.evictions + s2.evictions
        )
    }
}

// MARK: - Factory

public enum CacheFactory {
    public static func memory<V: Codable & Sendable>(capacity: Int = 256) -> MemoryCache<V> {
        MemoryCache(capacity: capacity)
    }

    public static func layered<V: Codable & Sendable>(
        l1Capacity: Int = 64,
        l2Capacity: Int = 512
    ) -> LayeredCache<V> {
        LayeredCache(l1Capacity: l1Capacity, l2Capacity: l2Capacity)
    }
}
